File: src/imguizmo/imguizmo.cpp
#include <util.hpp>

#include <imgui.h>
#include <ImGuizmo.h>
#include <cstdio>

// clang-format off
static auto const IMGUIZMO = bindings([]() {

bind_fn("ImGuizmo_BeginFrame", []() {
    ImGuizmo::BeginFrame();
});

bind_fn("ImGuizmo_SetDrawlist", []() {
    printf("[ImGuizmo C++] SetDrawlist\n");
    ImGuizmo::SetDrawlist();
    printf("[ImGuizmo C++] SetDrawlist done\n");
});

bind_fn("ImGuizmo_SetRect", [](float x, float y, float width, float height) {
    printf("[ImGuizmo C++] SetRect(%f, %f, %f, %f)\n", x, y, width, height);
    ImGuizmo::SetRect(x, y, width, height);
    printf("[ImGuizmo C++] SetRect done\n");
});

bind_fn("ImGuizmo_Enable", [](bool enable) {
    printf("[ImGuizmo C++] Enable(%d)\n", enable);
    ImGuizmo::Enable(enable);
    printf("[ImGuizmo C++] Enable done\n");
});

bind_fn("ImGuizmo_SetGizmoSizeClipSpace", [](float value) {
    printf("[ImGuizmo C++] SetGizmoSizeClipSpace(%f)\n", value);
    ImGuizmo::SetGizmoSizeClipSpace(value);
    printf("[ImGuizmo C++] SetGizmoSizeClipSpace done\n");
});

bind_fn("ImGuizmo_IsOver", []() {
    return ImGuizmo::IsOver();
});

bind_fn("ImGuizmo_IsUsing", []() {
    return ImGuizmo::IsUsing();
});

bind_fn("ImGuizmo_SetOrthographic", [](bool isOrthographic) {
    ImGuizmo::SetOrthographic(isOrthographic);
});

bind_fn("ImGuizmo_Manipulate", [](
    js_val view,
    js_val projection,
    int operation,
    int mode,
    js_val matrix,
    js_val snap
) {
    printf("[ImGuizmo C++] Manipulate op=%d mode=%d\n", operation, mode);
    auto param_view = get_array_param<float, 16>(view);
    auto param_projection = get_array_param<float, 16>(projection);
    auto param_matrix = get_array_param<float, 16>(matrix);
    auto param_snap = get_array_param<float, 3>(snap);
    printf("[ImGuizmo C++] arrays parsed: view=%p proj=%p mat=%p snap=%p\n",
        (void*)param_view.ptr, (void*)param_projection.ptr, (void*)param_matrix.ptr, (void*)param_snap.ptr);

    bool modified = ImGuizmo::Manipulate(
        param_view.ptr,
        param_projection.ptr,
        static_cast<ImGuizmo::OPERATION>(operation),
        static_cast<ImGuizmo::MODE>(mode),
        param_matrix.ptr,
        nullptr, // deltaMatrix
        param_snap.ptr,
        nullptr, // localBounds
        nullptr  // boundsSnap
    );

    if (modified) {
        write_back_array_param(param_matrix, matrix);
    }

    return modified;
});

bind_fn("ImGuizmo_DecomposeMatrixToComponents", [](js_val matrix) {
    auto param_matrix = get_array_param<float, 16>(matrix);
    float translation[3], rotation[3], scale[3];

    ImGuizmo::DecomposeMatrixToComponents(
        param_matrix.ptr,
        translation,
        rotation,
        scale
    );

    auto result = js_val::object();
    auto t = js_val::array();
    auto r = js_val::array();
    auto s = js_val::array();
    for (int i = 0; i < 3; i++) {
        t.set(i, translation[i]);
        r.set(i, rotation[i]);
        s.set(i, scale[i]);
    }
    result.set("translation", t);
    result.set("rotation", r);
    result.set("scale", s);
    return result;
});

bind_fn("ImGuizmo_RecomposeMatrixFromComponents", [](
    js_val translation,
    js_val rotation,
    js_val scale
) {
    auto param_t = get_array_param<float, 3>(translation);
    auto param_r = get_array_param<float, 3>(rotation);
    auto param_s = get_array_param<float, 3>(scale);
    float matrix[16];

    ImGuizmo::RecomposeMatrixFromComponents(
        param_t.ptr,
        param_r.ptr,
        param_s.ptr,
        matrix
    );

    auto result = js_val::array();
    for (int i = 0; i < 16; i++) {
        result.set(i, matrix[i]);
    }
    return result;
});

});
// clang-format on

File: src/imguizmo/imguizmo.ts
// @ts-nocheck
import { Mod } from "./core.js";

export const ImGuizmo = {
    OPERATION: {
        TRANSLATE_X: 1 << 0,
        TRANSLATE_Y: 1 << 1,
        TRANSLATE_Z: 1 << 2,
        ROTATE_X: 1 << 3,
        ROTATE_Y: 1 << 4,
        ROTATE_Z: 1 << 5,
        ROTATE_SCREEN: 1 << 6,
        SCALE_X: 1 << 7,
        SCALE_Y: 1 << 8,
        SCALE_Z: 1 << 9,
        BOUNDS: 1 << 10,
        SCALE_XU: 1 << 11,
        SCALE_YU: 1 << 12,
        SCALE_ZU: 1 << 13,
        TRANSLATE: (1 << 0) | (1 << 1) | (1 << 2),
        ROTATE: (1 << 3) | (1 << 4) | (1 << 5) | (1 << 6),
        SCALE: (1 << 7) | (1 << 8) | (1 << 9),
        SCALEU: (1 << 11) | (1 << 12) | (1 << 13),
        UNIVERSAL: (1 << 0) | (1 << 1) | (1 << 2) | (1 << 3) | (1 << 4) | (1 << 5) | (1 << 6) | (1 << 11) | (1 << 12) | (1 << 13),
    },

    MODE: {
        LOCAL: 0,
        WORLD: 1,
    },

    BeginFrame(): void {
        Mod.export.ImGuizmo_BeginFrame();
    },

    SetDrawlist(): void {
        Mod.export.ImGuizmo_SetDrawlist();
    },

    SetRect(x: number, y: number, width: number, height: number): void {
        Mod.export.ImGuizmo_SetRect(x, y, width, height);
    },

    Enable(enable: boolean): void {
        Mod.export.ImGuizmo_Enable(enable);
    },

    SetGizmoSizeClipSpace(value: number): void {
        Mod.export.ImGuizmo_SetGizmoSizeClipSpace(value);
    },

    IsOver(): boolean {
        return Mod.export.ImGuizmo_IsOver();
    },

    IsUsing(): boolean {
        return Mod.export.ImGuizmo_IsUsing();
    },

    SetOrthographic(isOrthographic: boolean): void {
        Mod.export.ImGuizmo_SetOrthographic(isOrthographic);
    },

    Manipulate(
        view: number[],
        projection: number[],
        operation: number,
        mode: number,
        matrix: number[],
        snap?: number[] | null,
    ): boolean {
        return Mod.export.ImGuizmo_Manipulate(
            view,
            projection,
            operation,
            mode,
            matrix,
            snap ?? null,
        );
    },

    DecomposeMatrixToComponents(matrix: number[]): {
        translation: number[];
        rotation: number[];
        scale: number[];
    } {
        return Mod.export.ImGuizmo_DecomposeMatrixToComponents(matrix);
    },

    RecomposeMatrixFromComponents(
        translation: number[],
        rotation: number[],
        scale: number[],
    ): number[] {
        return Mod.export.ImGuizmo_RecomposeMatrixFromComponents(
            translation,
            rotation,
            scale,
        );
    },
};
